nucypher/characters/lawful.py

```diff
@@ -678,28 +663,104 @@ def work_orders_for_capsules(self,
                 # TODO: Presently, the order here is haphazard .  Do we want to do the complete or incomplete specifically first? NRN
                 break
 
-        if incomplete_work_orders == OrderedDict():
-            self.log.warn("No new WorkOrders created. Try calling this with different parameters.")  # TODO: Clearer instructions.  NRN
+        if len(incomplete_work_orders) == 0:
+            self.log.warn(
+                "No new WorkOrders created.  Try calling this with different parameters.")  # TODO: Clearer instructions.  NRN
 
         return incomplete_work_orders, complete_work_orders
 
-    def get_reencrypted_cfrags(self, work_order, retain_cfrags=False):
-        if work_order.completed:
-            raise TypeError("This WorkOrder is already complete; "
-                            "if you want Ursula to perform additional service, make a new WorkOrder.")
-
-        cfrags_and_signatures = self.network_middleware.reencrypt(work_order)
-        cfrags = work_order.complete(cfrags_and_signatures)
-        self._completed_work_orders.save_work_order(work_order, as_replete=retain_cfrags)
-
-        return cfrags
-
     def join_policy(self, label, alice_verifying_key, node_list=None, block=False):
         if node_list:
             self._node_ids_to_learn_about_immediately.update(node_list)
         treasure_map = self.get_treasure_map(alice_verifying_key, label)
         self.follow_treasure_map(treasure_map=treasure_map, block=block)
 
+    def _fix_message_kits(self, message_kits, enrico, policy_encrypting_key, use_attached_cfrags):
```

Comment:
...fix?

Reply:
The logic that I put here checks if the `sender` attribute is correct and changes it under some conditions. So, I think, "fix" is justified, but it can be made more specific, "ensure_correct_sender" or something?

It may work better as a method of `UmbralMessageKit`, but all the classes in `characters/control/specifications` are very barebones, so I'm not sure it will fit there.

Of course, ideally I'd get rid of mutation completely. But this PR just tries to extract existing pieces of logic into functions.

Reply:
```suggestion
    def _ensure_correct_sender(self, message_kits, enrico, policy_encrypting_key, use_attached_cfrags):
```

Reply:
The actual UmbralMessagekit class is here:
https://github.com/nucypher/nucypher/blob/main/nucypher/crypto/kits.py#L88
Maybe it would be better as a method on this class.

Reply:
Making it a method there would require importing `Enrico` from `lawful.py`, which will lead to a circular dependency... For now just made it a standalone function with a more specific name.